File: rs/bubbleSort.rs
use std::io;

fn main() {
    let mut s = String::new();
    io::stdin().read_line(&mut s);
    let mut xs: Vec<i64> = s.trim().split(' ').map(|x| x.parse().unwrap()).collect();

    let mut count = 0;
    let mut flag = true;
    while flag {
        flag = false;
        for i in 0..(xs.len() - 1) {
            if xs[i] > xs[i + 1] {
                xs.swap(i, i + 1);
                flag = true;
                count += 1;
            }
        }
    }

    for x in xs {
        print!("{} ", x);
    }
    println!("");

    println!("{}", count);
}

File: scala/bubbleSort.scala
import scala.io.StdIn._

object Main {
    var count = 0

    def main(args: Array[String]) {
        val xs = readLine().split(" ").map {_.toInt}

        bubbleSort(xs.toList).foreach {x => print(x + " ")}
        println("")
        println(count)
    }

    def bubbleSort(xs: List[Int]): List[Int] = {
        innerBSort(xs) match {
            case (res, true) => bubbleSort(res)
            case (res, _) => res
        }
    }

    def innerBSort(xs: List[Int]): (List[Int], Boolean) = {
        xs match {
            case a :: b :: cs if a > b =>
                val (res, _) = innerBSort (a :: cs)
                count += 1
                (b :: res, true)
            case a :: b :: cs =>
                val (res, f) = innerBSort (b :: cs)
                (a :: res, f)
            case _ => (xs, false)
        }
    }
}
